blue_hardware_interface/src/blue_hardware_interface.cpp

```diff
@@ -68,9 +71,90 @@ BlueHW::BlueHW(ros::NodeHandle &nh) : nh_(nh) {
       this);
 }
 
+void BlueHW::doSwitch(const std::list<hardware_interface::ControllerInfo>& start_list,
+                      const std::list<hardware_interface::ControllerInfo>& stop_list) {
+
+  // TODO read these mappings in as a parameter
+  std::vector<std::string> joint_hardware_effort_controllers{
+    "effort_controllers/JointTrajectoryController",
+    "effort_controllers/JointGroupEffortController",
+    "blue_controllers/BlueJointGroupPositionController",
+    "blue_controllers/BlueJointGroupCTC",
+  };
+  std::vector<std::string> joint_hardware_position_controllers{
+    "position_controllers/JointGroupPositionController",
+  };
+
+  std::vector<std::string> gripper_hardware_effort_controllers{
+    "effort_controllers/GripperActionController",
+    "effort_controllers/JointGroupEffortController",
+  };
+  std::vector<std::string> gripper_hardware_position_controllers{
+  };
+
+  // Acquire the motor_driver lock
+  std::lock_guard<std::mutex> lock(motor_driver_mutex_);
+
+
+  for (std::list<hardware_interface::ControllerInfo>::const_iterator it=stop_list.begin(); it != stop_list.end(); ++it) {
+    // If stopping a joint position hardware control mode, then switch back to current control mode for gravity comp
+    if (std::find(
+          std::begin(joint_hardware_position_controllers),
+          std::end(joint_hardware_position_controllers),
+          it->type) != std::end(joint_hardware_position_controllers)) {
+
+      // Change the control mode for all non gripper motors
+      for (int i = 0; i < kinematics_.getJointCount() - 1; i++)
+        motor_driver_.setControlMode(i, blue_hardware_drivers::COMM_CTRL_MODE_CURRENT);
```

Comment:
there seems to be an indexing bug everywhere you call `setControlMode` -- `setControlMode` expects a board ID (`comm_id_t`) for the first parameter, not a board index

these are unfortunately 1-indexed, since we set 0 to be the broadcast address (maybe this should be changed to 255? that sounds more standard...)

as a result, the first iteration of this loop will set the control mode for all boards, including the gripper :(


I think what you're looking for is:
```
      for (int i = 0; i < kinematics_.getJointCount() - 1; i++)
        motor_driver_.setControlMode(board_ids_[i], blue_hardware_drivers::COMM_CTRL_MODE_CURRENT);
```